silence/silence.go

```diff
@@ -487,6 +497,29 @@ var ValidateMatcher = func(m *pb.Matcher) error {
 	return nil
 }
 
+// validateUTF8Matcher validates the matcher against the UTF-8 rules.
+func validateUTF8Matcher(m *pb.Matcher) error {
+	if !utf8.ValidString(m.Name) {
+		return fmt.Errorf("invalid label name %q", m.Name)
+	}
+	switch m.Type {
+	case pb.Matcher_EQUAL, pb.Matcher_NOT_EQUAL:
+		if !utf8.ValidString(m.Pattern) {
+			return fmt.Errorf("invalid label value %q", m.Pattern)
+		}
+	case pb.Matcher_REGEXP, pb.Matcher_NOT_REGEXP:
+		if !utf8.ValidString(m.Pattern) {
+			return fmt.Errorf("invalid regular expression %q", m.Pattern)
+		}
+		if _, err := regexp.Compile(m.Pattern); err != nil {
+			return fmt.Errorf("invalid regular expression %q: %s", m.Pattern, err)
+		}
+	default:
+		return fmt.Errorf("unknown matcher type %q", m.Type)
+	}
+	return nil
+}
+
```

Comment:
I have my reservations when it comes to this code - to me, this feels wrong. We're accepting matches as user input from the silences API in a structured format:

```
 "matchers": [
    {
      "name": "string",
      "value": "string",
      "isRegex": true,
      "isEqual": true
    }
  ],
```

In my opinion, this validation should happen at an API level and ideally make the input matchers go through the parser, by taking each parser and constructing a string that would you give you the struct directly. 

My main worry is that right now validation for entities is spreadout in multiple layers which makes it difficult to track who validates what.

Not something that we need to solve right now, but it's worth thinking about.